### PR TITLE
New api call

### PR DESCRIPTION
**Pull request by allthatilk**
Created a new API call to provide a list of dataset dictionaries with their resources in order to test `last_successful_fetch` without worrying about timeout whilst maintaining backwards compatibility for other Datastore API calls.

### CODE REVIEW

**Comment by andylolz** [comment]
@allthatilk: I think the join there just skips datasets with zero resources… The resources get lazy loaded, so it doesn’t speed things up at all.

Related to that, I wonder if this needs a `db.subqueryload` to avoid the query taking a reeeeally long time. I.e.:

```python
dataset_resources = db.session.query(Dataset).options(db.subqueryload(Dataset.resources))
```

**Comment by allthatilk** [comment]
@andylolz Thanks for the suggestion. I'm afraid my knowledge of SQLAlchemy is limited and even after research and debugging I can't get your suggestion to work. If you have a working alternative a PR would be appreciated.

As for speed, this is more of a hacky solution to Travis timing out before completing the website test we made for `last_successful_fetch` as we need a way to monitor the datastore's update process (and catch problems on the publisher end). Basically, as long as it's only doing one slow query, that should (hopefully) be good enough, rather than the only other current option of making a loop of thousands of querys.

**Comment by andylolz** [comment]
> Basically, as long as it's only doing one slow query, that should (hopefully) be good enough, rather than the only other current option of making a loop of thousands of querys.

So I don’t know if you know this already (apologies if you do!) but you can run a shell with:

```shell
$ cd IATI-Datastore
$ source env/bin/activate
$ iati shell
```

Then in the shell, you can set [this very magical and useful setting](http://flask-sqlalchemy.pocoo.org/2.1/config/):

```python
from flask_sqlalchemy import get_debug_queries

app.config['SQLALCHEMY_RECORD_QUERIES'] = True
```

…to store all the queries that SQLAlchemy is running. Super helpful for debugging! (NB `app` is already set in the context, so you don’t have to `import` anything extra to do this.)

Then if you do some necessary imports:

```python
from iatilib import db
from iatilib.model import Dataset
```

…and then run your code: https://github.com/IATI/IATI-Datastore/blob/52c8b2a51a7542006c650557ae7771fd2c80c278/iati_datastore/iatilib/frontend/api1.py#L74-L86

…and then:

```python
len(get_debug_queries())
```

…you’ll see that even though it might look like a single query, behind the scenes SQLAlchemy is doing loooots of queries (approx 10,000). So if you don’t explicitly tell SQLAlchemy to load relationships ahead of time, it will likely be just as slow, and Travis will still timeout 😭 

> If you have a working alternative a PR would be appreciated.

So here’s the patch I’m suggesting, but note that while it improves matters, it only halves the number of queries. That’s because `dataset.resources[0].activities.count()` still runs lots of queries.

```patch
diff --git a/iati_datastore/iatilib/frontend/api1.py b/iati_datastore/iatilib/frontend/api1.py
index 8e1dc79..1f5319d 100644
--- a/iati_datastore/iatilib/frontend/api1.py
+++ b/iati_datastore/iatilib/frontend/api1.py
@@ -71,10 +71,12 @@ def about_dataset(dataset):
 @api.route('/about/datasets/nest')
 def nest_about_dataset():
     """Output a JSON formatted list of dataset dictionaries containing their resource details."""
-    dataset_resources = db.session.query(Dataset).join(Dataset.resources)
+    dataset_resources = db.session.query(Dataset).options(db.subqueryload(Dataset.resources))
     datasets = dict()
 
     for dataset in dataset_resources:
+        if len(dataset.resources) == 0:
+            continue
         resources = {
             'url': dataset.resources[0].url,
             'last_fetch': dataset.resources[0].last_fetch.isoformat() if dataset.resources[0].last_fetch else None,
```

I’m sure it’s possible to do better, but I’m also not very familiar with SQLAlchemy.

**Comment by allthatilk** [comment]
Ooo! Thanks for the debugging tips! Much appreciated 😄  Your solution works great! We'll see how Travis likes it now, if it's still too slow we might have to change the approach for the call and have it as a basic fetch status call instead so we don't have to load the number of datasets.

**Comment by hayfield** [comment]
@allthatilk Is this complete and ready to review?

(the Datastore doesn't yet appear to have a `complete` label)

File: iati_datastore/iatilib/frontend/api1.py [hunk 1]
```diff
@@ -67,6 +67,32 @@ def about_dataset(dataset):
             resources=resources,
     )
 
+
+@api.route('/about/datasets/nest')
```
Comment (dalepotter) [inline iati_datastore/iatilib/frontend/api1.py]: Looking at this again, I'm wondering about the name... Would it be clearer if the API call was `about/datasets/fetch_status` ??

Comment (allthatilk) [reply iati_datastore/iatilib/frontend/api1.py]: done

File: docs/api/error.rst [hunk 2]
```diff
@@ -25,6 +22,8 @@ To retrieve metadata on the datasets known to the datastore.
       -  ``last_parsed`` - last date resource was successfully parsed
       -  ``num_of_activities`` number of activities datastore has
          successfully parsed and stored
+-   ``/api/1/about/datasets/nest`` retrieves the list of datasets currently
```
Comment (hayfield) [inline docs/api/error.rst]: Name now changes

File: iati_datastore/iatilib/frontend/api1.py [hunk 3]
```diff
@@ -67,6 +67,32 @@ def about_dataset(dataset):
             resources=resources,
     )
 
+
+@api.route('/about/datasets/fetch_status')
+def nest_about_dataset():
+    """Output a JSON formatted list of dataset dictionaries containing their resource details.
+    This is an experimental API call and not intended for general use.
```
Comment (hayfield) [inline iati_datastore/iatilib/frontend/api1.py]: Suggest putting this within a `Warning` scope.

File: iati_datastore/iatilib/test/test_api.py [hunk 4]
```diff
@@ -23,6 +23,14 @@ def test_about_http(self):
         resp = self.client.get('/api/1/about')
         self.assertEquals(200, resp.status_code)
 
+class TestAboutDatasets(ClientTestCase):
+    def test_about_datasets_nest(self):
+        """Check that the `about/datasets/nest` page has a 200 response and contains expected data."""
+        resp = self.client.get('/api/1/about/datasets/nest')
```
Comment (hayfield) [inline iati_datastore/iatilib/test/test_api.py]: `nest` -> new name

File: iati_datastore/iatilib/test/test_api.py [hunk 5]
```diff
@@ -23,6 +23,14 @@ def test_about_http(self):
         resp = self.client.get('/api/1/about')
         self.assertEquals(200, resp.status_code)
 
+class TestAboutDatasets(ClientTestCase):
+    def test_about_datasets_nest(self):
```
Comment (hayfield) [inline iati_datastore/iatilib/test/test_api.py]: `nest` -> new name

File: docs/api/error.rst (hunk 2 above)
Comment (allthatilk) [reply docs/api/error.rst]: actually if we're not documenting it this should go entirely

File: iati_datastore/iatilib/frontend/api1.py (hunk 3 above)
Comment (allthatilk) [reply iati_datastore/iatilib/frontend/api1.py]: done

File: iati_datastore/iatilib/test/test_api.py (hunk 5 above)
Comment (allthatilk) [reply iati_datastore/iatilib/test/test_api.py]: done

**Pull request merged**